coreruleset.go

```diff
@@ -4,16 +4,60 @@ import (
 	"embed"
 	"io/fs"
 	"log"
+	"strings"
 )
 
+// subFS is the interface that contains the methods that are used by
+// `fs.subFS`
+type subFS interface {
+	Open(name string) (fs.File, error)
+	ReadDir(name string) ([]fs.DirEntry, error)
+	ReadFile(name string) ([]byte, error)
+	Glob(pattern string) ([]string, error)
+}
+
 //go:embed rules/*
 var wFS embed.FS
 var FS fs.FS
 
 func init() {
-	var err error
-	FS, err = fs.Sub(wFS, "rules")
+	rulesFS, err := fs.Sub(wFS, "rules")
 	if err != nil {
 		log.Fatal(err)
 	}
+	FS = wrapFS{rulesFS.(subFS)}
+}
+
+// wrapFS is a wrapper around a subFS that removes the absolute path prefix
+// from the `name` argument in the ReadFile method.
+// The rationale is that the coraza parser when asked to load the files
+// invoked by `Include` will prepend the dir path of the parent file.
+// For example if you aim to load directives from `my-file.conf which contains`:
+//
+//	`Include @owasp_crs/REQUEST-911-METHOD-ENFORCEMENT.conf`
+//
+// the way the parser will load the REQUEST-911-METHOD-ENFORCEMENT.conf file
+// will depend on how the config was passed on the `WithDirectivesFromFile` method:
+//   - `my-file.conf`: the dir is "" and hence it will call the RootFS to load
+//     `@owasp_crs/REQUEST-911-METHOD-ENFORCEMENT.conf`
+//   - `../my-file.conf`: the dir is `..` and hence it will call the RootFS to load
+//     `../@owasp_crs/REQUEST-911-METHOD-ENFORCEMENT.conf` which can't be resolved
```

Comment:
Thanks for the examples, now I get it.

I think this approach currently puts too much meaning on the single character `@`, a user's path may have that character and wouldn't resolve correctly anymore AFAICT. Also, it's perfectly fine for there to actually be a `../@owasp_crs` folder on the FS and that should be usable

I would

- Change to search for the full string `@owasp_crs`
- First `ReadFile` with the original path and only if it fails fallback to the mangled one

More generally, we may also want to reconsider the `@` scheme for using real absolute paths - a lot of the problem seems to come from treating a relative path as effectively an absolute path within the embedded file system. So perhaps we should just have users use an absolute path, `/embedded/owasp_crs` or just `/owasp_crs`? We would still need the wrapper to remove the leading slash or prefix before reading, but it seems to be a more direct approach than looking for a magic string within the path